precompiles/balances-erc20/src/lib.rs

```diff
@@ -0,0 +1,412 @@
+pub type ApprovesStorage<Runtime, Instance> = StorageDoubleMap<
+	<Instance as InstanceToPrefix>::ApprovesPrefix,
+	Blake2_128Concat,
+	<Runtime as frame_system::Config>::AccountId,
+	Blake2_128Concat,
+	<Runtime as frame_system::Config>::AccountId,
+	BalanceOf<Runtime, Instance>,
+>;
+
+#[precompile_utils::generate_function_selector]
+#[derive(Debug, PartialEq, num_enum::TryFromPrimitive, num_enum::IntoPrimitive)]
+pub enum Action {
+	TotalSupply = "totalSupply()",
+	BalanceOf = "balanceOf(address)",
+	Allowance = "allowance(address,address)",
+	Transfer = "transfer(address,uint256)",
+	Approve = "approve(address,uint256)",
+	TransferFrom = "transferFrom(address,address,uint256)",
+}
+
+/// Precompile exposing a pallet_balance as an ERC20.
+/// Multiple precompiles can support instances of pallet_balance.
+/// The precompile uses an additional storage to store approvals.
+pub struct Erc20BalancesPrecompile<Runtime, Instance: 'static = ()>(
+	PhantomData<(Runtime, Instance)>,
+);
+
+impl<Runtime, Instance> Precompile for Erc20BalancesPrecompile<Runtime, Instance>
+where
+	Instance: InstanceToPrefix + 'static,
+	Runtime: pallet_balances::Config<Instance> + pallet_evm::Config,
+	Runtime::Call: Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+	Runtime::Call: From<pallet_balances::Call<Runtime, Instance>>,
+	<Runtime::Call as Dispatchable>::Origin: From<Option<Runtime::AccountId>>,
+	BalanceOf<Runtime, Instance>: TryFrom<U256> + Into<U256>,
+{
+	fn execute(
+		input: &[u8], //Reminder this is big-endian
+		target_gas: Option<u64>,
+		context: &Context,
+	) -> Result<PrecompileOutput, ExitError> {
+		let mut input = EvmDataReader::new(input);
+
+		match &input.read_selector()? {
+			Action::TotalSupply => Self::total_supply(input, target_gas),
+			Action::BalanceOf => Self::balance_of(input, target_gas),
+			Action::Allowance => Self::allowance(input, target_gas),
+			Action::Approve => Self::approve(input, target_gas, context),
+			Action::Transfer => Self::transfer(input, target_gas, context),
+			Action::TransferFrom => Self::transfer_from(input, target_gas, context),
+		}
+	}
+}
+
+impl<Runtime, Instance> Erc20BalancesPrecompile<Runtime, Instance>
+where
+	Instance: InstanceToPrefix + 'static,
+	Runtime: pallet_balances::Config<Instance> + pallet_evm::Config,
+	Runtime::Call: Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+	Runtime::Call: From<pallet_balances::Call<Runtime, Instance>>,
+	<Runtime::Call as Dispatchable>::Origin: From<Option<Runtime::AccountId>>,
+	BalanceOf<Runtime, Instance>: TryFrom<U256> + Into<U256>,
+{
+	fn total_supply(input: EvmDataReader, target_gas: Option<u64>) -> EvmResult<PrecompileOutput> {
+		let mut gasometer = Gasometer::new(target_gas);
+		gasometer.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+
+		// Parse input.
+		input.expect_arguments(0)?;
+
+		// Fetch info.
+		let amount: U256 = pallet_balances::Pallet::<Runtime, Instance>::total_issuance().into();
```

Comment:
Why is this inefficient ? Also, how to do the conversion by hand if we're generic over the Currency type ?

Reply:
Alan had observed that U256::from had performance problems, and that minimizing them offers a significant gain, I can't say why, maybe the fact that the cpu is optimized to handle smaller integers.  
And indeed one has to choose between performance or genericity.

Reply:
For context, I said that for the EVM where every opcode was computing the U256::FROM a static number, which was bad because we could pre-compute it.
In this case, I'm not sure the performance impact